controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,349 @@
+			continue
+		}
+
+		apiGroups, ok := newKind["apiGroups"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		kindsInKinds, ok := newKind["kinds"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		for _, apiGroup := range apiGroups {
+			for _, kind := range kindsInKinds {
+				version, err := r.getAPIVersion(kind.(string), apiGroup.(string), false)
+				if err != nil {
+					r.Log.V(1).Info("getAPIVersion has error but continue")
+
+					finalErr = err
+				}
+
+				syncOnlys = append(syncOnlys, v1alpha1.SyncOnlyEntry{
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kind.(string),
+				})
+			}
+		}
+	}
+
+	return &syncOnlys, finalErr
+}
+
+// Find the version that the constraint does not provided.
+// Constraint only provide kind and apiGroup. However the config resource need version
+func (r *ConstraintStatusReconciler) getAPIVersion(kind string, apiGroup string, skipRefresh bool) (string, error) {
+	// Cool time(10 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+
+		// The discovery is just refeshed so skip another refesh
+		skipRefresh = true
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot parse the group and version in getApiVersion ", "GroupVersion:", resc.GroupVersion)
+
+			continue
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	if !skipRefresh {
+		// Get new discoveryInfo, when any resource is not found
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+		// Retry one more time after refresh the discovery
+		return r.getAPIVersion(kind, apiGroup, true)
+	}
+
+	return "", ErrNotFoundDiscovery
+}
+
+// Retrieve all groups and versions to add in config sync
+// Constraints present only kind and group so this function helps to find the version
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
+	discoveryClient := r.ClientSet.Discovery()
+
+	apiList, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return err
+	}
+
+	// Save fetched discovery at apiResourceList
+	r.apiResourceList = apiList
+
+	return nil
+}
+
+func (r *ConstraintStatusReconciler) getUniqSyncOnly() []v1alpha1.SyncOnlyEntry {
+	syncOnlySet := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for unique filtering
+	for _, syncEntries := range r.ConstraintToSyncOnly {
+		for _, entry := range syncEntries {
+			syncOnlySet[entry] = true
+		}
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
```

Comment:
This sets up the exact size of the array in memory so Go doesn't ever need to allocate a new array when calling `append`.
```suggestion
	syncOnlys := make([]v1alpha1.SyncOnlyEntry, 0, len(syncOnlySet))
```